File: src/main/java/outland/emr/tracking/controllers/BeaconController.java
package outland.emr.tracking.controllers;

import outland.emr.tracking.dtos.BeaconDTO;
import outland.emr.tracking.logic.BeaconLogic;
import outland.emr.tracking.mappers.Mapper;
import outland.emr.tracking.models.BeaconType;
import outland.emr.tracking.models.socket.Stream;
import outland.emr.tracking.websockets.TrackingSocket;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.io.IOException;
import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/beacon")
public class BeaconController {
    @Autowired
    private final BeaconLogic beaconLogic;
    @Autowired
    private final Mapper mapper;
    @Autowired
    private final TrackingSocket trackingSocket;

    public BeaconController(BeaconLogic beaconLogic, Mapper mapper, TrackingSocket trackingSocket) {
        this.beaconLogic = beaconLogic;
        this.mapper = mapper;
        this.trackingSocket = trackingSocket;
    }

    @GetMapping("/available")
    public List<BeaconDTO> getAvailableBeacons() {
        return beaconLogic.getAvailable()
                .stream()
                .filter(beacon -> beacon.getType().equals(BeaconType.PATIENT))
                .map(beacon -> mapper.getMapper().map(beacon, BeaconDTO.class))
                .collect(Collectors.toList());
    }

    @PutMapping("/associate")
    public void associate(@RequestParam String patientId, @RequestParam String uniqueId) {
        Stream stream = beaconLogic.associate(patientId, uniqueId);
        if (stream != null) {
            try {
                trackingSocket.emitBeaconUpdate(stream);
            } catch (IOException ignored) { }
        }
    }

    @PutMapping("/deassociate")
    public void deassociate(@RequestParam String beaconLabel) {
        Stream stream = beaconLogic.disassociate(beaconLabel);
        if (stream != null) {
            try {
                trackingSocket.emitBeaconDetachment(stream);
            } catch (IOException ignored) { }
        }
    }
}

File: src/main/kotlin/com/emr/tracking/controller/BeaconController.kt
package com.emr.tracking.controller

import com.emr.tracking.manager.StreamManager
import com.emr.tracking.model.*
import com.emr.tracking.repository.MongoBeaconRepository
import com.emr.tracking.repository.MongoPatientBeaconRegistry
import com.emr.tracking.repository.MongoPatientRepository
import com.emr.tracking.websocket.TrackingSocket
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime
import java.util.*

@RestController
@RequestMapping("/beacon")
class BeaconController(
    private val mongoPatientRepository: MongoPatientRepository,
    private val mongoBeaconRepository: MongoBeaconRepository,
    private val mongoPatientBeaconRegistry: MongoPatientBeaconRegistry,
    private val trackingSocket: TrackingSocket,
    private val streamManager: StreamManager
) {

    @GetMapping("/available")
    fun getAll(): List<Beacon> {
        val lockedBeacons = mongoPatientBeaconRegistry.findAll().filter { it.active }.map { it.beaconId }
        return mongoBeaconRepository.findAll().filter { !lockedBeacons.contains(it.mac) }
    }

    @GetMapping("/activated")
    fun getActivated(): List<Beacon> {
        val lockedBeaconIds = mongoPatientBeaconRegistry.findAll().filter { it.active }.map { it.beaconId }
        return mongoBeaconRepository.findAll().filter { lockedBeaconIds.contains(it.mac) }
    }

    @PutMapping("/associate")
    fun associate(@RequestParam patientId: String, @RequestParam uniqueId: String): Patient {
        val patientPromise = mongoPatientRepository.findById(patientId)
        if (patientPromise.isEmpty) {
            throw Error("Patient not found")
        }
        val beacon: Beacon = mongoBeaconRepository.findAll().first { it.uniqueId == uniqueId }
            ?: throw Error("Beacon not found")
        val patient = patientPromise.get()

        val oldRegistry = mongoPatientBeaconRegistry.findAll().firstOrNull { it.active && it.patientId == patientId }
        val patientBeaconRegistry = PatientBeaconRegistry(UUID.randomUUID().toString(), patientId, beacon.mac, true, LocalDateTime.now(), LocalDateTime.now())
        mongoPatientBeaconRegistry.save(patientBeaconRegistry)

        if (oldRegistry != null) {
            oldRegistry.active = false
            mongoPatientBeaconRegistry.save(oldRegistry)
        }

        val stream = streamManager.createStreamSocketFor(patient, patientBeaconRegistry)
        if (stream != null) {
            trackingSocket.emitBeaconUpdate(stream)
        }

        patient.trackingDeviceId = beacon.uniqueId
        return mongoPatientRepository.save(patient)
    }

    @PutMapping("/deassociate")
    fun deassociate(@RequestParam patientId: String): Patient {
        val patientPromise = mongoPatientRepository.findById(patientId)
        if (patientPromise.isEmpty) {
            throw Error("Patient not found")
        }
        val patient = patientPromise.get()

        val patientBeaconRegistry = mongoPatientBeaconRegistry.findAll().first { it.active && it.patientId == patientId }

        val streamSocket = streamManager.createStreamSocketForPatient(patient)
        if (streamSocket != null) {
            trackingSocket.emitBeaconDetachment(streamSocket)
        }

        patientBeaconRegistry.active = false
        patientBeaconRegistry.updatedAt = LocalDateTime.now()

        mongoPatientBeaconRegistry.save(patientBeaconRegistry)

        patient.trackingDeviceId = ""
        return mongoPatientRepository.save(patient)
    }

    @PostMapping
    fun addDevices(@RequestBody devices: List<Beacon>): List<Beacon> {
        return mongoBeaconRepository.saveAll(devices)
    }
}